cmd/routed-eni-cni-plugin/driver/driver.go

```diff
@@ -267,354 +271,358 @@ func WaitForAddressesToBeStable(ifName string, timeout time.Duration) error {
+		containerAddr = v6Addr
+	}
 
-	var addrHostAddr *net.IPNet
-	//We only support either v4 or v6 modes.
-	if v4Addr != nil && v4Addr.IP.To4() != nil {
-		addrHostAddr = &net.IPNet{
-			IP:   v4Addr.IP,
-			Mask: net.CIDRMask(32, 32)}
-	} else if v6Addr != nil && v6Addr.IP.To16() != nil {
-		addrHostAddr = &net.IPNet{
-			IP:   v6Addr.IP,
-			Mask: net.CIDRMask(128, 128)}
+	rtTable := unix.RT_TABLE_MAIN
+	if deviceNumber > 0 {
+		rtTable = deviceNumber + 1
 	}
+	if err := n.setupIPBasedContainerRouteRules(hostVeth, containerAddr, rtTable, log); err != nil {
+		return errors.Wrapf(err, "SetupPodNetwork: unable to setup IP based container routes and rules")
+	}
+	return nil
+}
 
-	// Add host route
-	route := netlink.Route{
-		LinkIndex: hostVeth.Attrs().Index,
-		Scope:     netlink.SCOPE_LINK,
-		Dst:       addrHostAddr}
+// TeardownPodNetwork cleanup ip rules
+func (n *linuxNetwork) TeardownPodNetwork(containerAddr *net.IPNet, deviceNumber int, log logger.Logger) error {
+	log.Debugf("TeardownPodNetwork: containerAddr=%s, deviceNumber=%d", containerAddr.String(), deviceNumber)
 
-	// Add or replace route
-	if err := netLink.RouteReplace(&route); err != nil {
-		return errors.Wrapf(err, "setupNS: unable to add or replace route entry for %s", route.Dst.IP.String())
+	rtTable := unix.RT_TABLE_MAIN
+	if deviceNumber > 0 {
+		rtTable = deviceNumber + 1
 	}
-	log.Debugf("Successfully set host route to be %s/0", route.Dst.IP.String())
+	if err := n.teardownIPBasedContainerRouteRules(containerAddr, rtTable, log); err != nil {
+		return errors.Wrapf(err, "TeardownPodNetwork: unable to teardown IP based container routes and rules")
+	}
+	return nil
+}
 
-	err = addContainerRule(netLink, true, addrHostAddr, mainRouteTable)
+// SetupBranchENIPodNetwork sets up the network ns for pods requesting its own security group
+func (n *linuxNetwork) SetupBranchENIPodNetwork(hostVethName string, contVethName string, netnsPath string, v4Addr *net.IPNet, v6Addr *net.IPNet,
+	vlanID int, eniMAC string, subnetGW string, parentIfIndex int, mtu int, podSGEnforcingMode sgpp.EnforcingMode, log logger.Logger) error {
+	log.Debugf("SetupBranchENIPodNetwork: hostVethName=%s, contVethName=%s, netnsPath=%s, v4Addr=%v, v6Addr=%v, vlanID=%d, eniMAC=%s, subnetGW=%s, parentIfIndex=%d, mtu=%d, podSGEnforcingMode=%v",
+		hostVethName, contVethName, netnsPath, v4Addr, v6Addr, vlanID, eniMAC, subnetGW, parentIfIndex, mtu, podSGEnforcingMode)
 
+	hostVeth, err := n.setupVeth(hostVethName, contVethName, netnsPath, v4Addr, v6Addr, mtu, log)
 	if err != nil {
-		log.Errorf("Failed to add toContainer rule for %s err=%v, ", addrHostAddr.String(), err)
-		return errors.Wrap(err, "setupNS network: failed to add toContainer")
+		return errors.Wrapf(err, "SetupBranchENIPodNetwork: failed to setup veth pair")
 	}
 
-	log.Infof("Added toContainer rule for %s", addrHostAddr.String())
+	// clean up any previous hostVeth ip rule recursively. (when pod with same name are recreated multiple times)
```

Comment:
Do we still need this? We're now using `prevResult` to drive the deletion and so (ideally) shouldn't run in to the issue anymore - #1705

Reply:
I keep this logic for safety purpose. but  i don't know why the original code explicitly delete this hostVeth but not the hostVlan rule here.

Reply:
Yes, if I remember correctly we can end up deleting the wrong Pod because of the race condition (i.e.,) by the time CNI makes an API Server call, the old pod might've been already deleted so it will not be able to derive the vlan id and upon retry it will get the pod spec of the new pod (with same name and ns) and uses it's vlan id to clean up the rules. Will result in CNI incorrectly deleting the vlan rules of the new pod instead of the old pod.

Reply:
Agree a comment here on why we were only deleting the `hostVeth` here would've been helpful.

Reply:
make sense, will add a comment here

Reply:
added a comment here.